File: tls-rds/ec2/go/main.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strings"

	"database/sql"
	"github.com/go-sql-driver/mysql"
)

const (
	rdsTlsCaCertPath = "/etc/ssl/certs/rds-combined-ca-bundle.pem"
)

func main() {
	dbUser := os.Getenv("DB_USER")
	dbPassword := os.Getenv("DB_PASSWORD")
	dbHost := os.Getenv("DB_HOST")
	dbName := os.Getenv("DB_NAME")
	if dbUser == "" {
		log.Fatal("Please supply DB_USER environment variable.")
	}
	if dbPassword == "" {
		log.Fatal("Please supply DB_PASSWORD environment variable.")
	}
	if dbHost == "" {
		log.Fatal("Please supply DB_HOST environment variable.")
	}
	if dbName == "" {
		log.Fatal("Please supply DB_NAME environment variable.")
	}
	pemFileData, err := ioutil.ReadFile(rdsTlsCaCertPath)
	if err != nil {
		log.Fatalf("Error opening RDS combined ca bundle: %v\n", err)
	}
	rootCaCertPool, err := x509.SystemCertPool()
	if err != nil {
		log.Fatalf("Error getting system cert pool: %v\n", err)
	}
	ok := rootCaCertPool.AppendCertsFromPEM(pemFileData)
	if !ok {
		log.Fatal("Failed to append certs from RDS combined ca bundle")
	}
	mysql.RegisterTLSConfig("custom", &tls.Config{
		RootCAs: rootCaCertPool,
	})
	dsn := fmt.Sprintf("%s:%s@tcp(%s)/%s?tls=custom", dbUser, dbPassword, dbHost, dbName)
	db, err := sql.Open("mysql", dsn)
	if err != nil {
		log.Fatalf("Error connecting to MySQL: %v\n", err)
	}
	defer db.Close()
	rows, err := db.Query("SELECT title FROM books LIMIT 1")
	if err != nil {
		msg := fmt.Sprintf("Error running SELECT statement: %v", err)
		log.Print(msg)
	} else {
		if rows.Next() {
			title := ""
			err := rows.Scan(&title)
			if err != nil {
				log.Printf("rows.Scan() error = %v\n", err)
				return
			}
			cols, err := rows.Columns()
			if err != nil {
				log.Printf("rows.Columns() error = %v\n", err)
			} else {
				log.Printf("Columns = %s\n", strings.Join(cols, ","))
				log.Printf("Book title = \"%s\"\n", title)
			}
		} else {
			log.Println("No results.")
		}
	}
}

// vim:noet

File: tls-rds/ec2/python/main.py
import os
import sys

import pymysql

_RDS_TLS_CA_CERT_PATH = "/etc/ssl/certs/rds-combined-ca-bundle.pem"

def _main():
    global _RDS_TLS_CA_CERT_PATH
    db_host = os.environ.get("DB_HOST", None)
    db_user = os.environ.get("DB_USER", None)
    db_password = os.environ.get("DB_PASSWORD", None)
    db_name = os.environ.get("DB_NAME", None)
    must_exit = False
    if db_host is None:
        must_exit = True
        print("Env var 'DB_HOST' is missing.")
    if db_user is None:
        must_exit = True
        print("Env var 'DB_USER' is missing.")
    if db_password is None:
        must_exit = True
        print("Env var 'DB_PASSWORD' is missing.")
    if db_name is None:
        must_exit = True
        print("Env var 'DB_NAME' is missing.")
    if must_exit:
        print("Exiting.")
        sys.exit(1)
    mysql_client = pymysql.connect(
        host=db_host,
        user=db_user,
        password=db_password,
        db=db_name,
        ssl={
            "ca": _RDS_TLS_CA_CERT_PATH,
        },
    )
    try:
        with mysql_client.cursor() as cursor:
            sql_query = "SELECT title FROM books LIMIT 1;"
            cursor.execute(sql_query)
            result = cursor.fetchone()
            print("Result = {}".format(result), flush=True)
    finally:
        mysql_client.close()

if __name__ == "__main__":
    _main()
